pkg/controller/volume/attach_detach_controller.go

```diff
@@ -106,30 +240,364 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+// the given name.
+// This method deep copies the PV object so the caller may use the returned
+// volume.Spec object without worrying about it mutating unexpectedly.
+func (adc *attachDetachController) getPVSpecFromCache(
+	name string,
+	pvcReadOnly bool,
+	expectedClaimUID types.UID) (*volume.Spec, error) {
+	pvObj, exists, err := adc.internalPVInformer.GetStore().Get(name)
+	if pvObj == nil || !exists || err != nil {
+		return nil, fmt.Errorf(
+			"failed to find PV %q in PVInformer cache. %v", name, err)
+	}
+
+	pv, ok := pvObj.(*api.PersistentVolume)
+	if ok || pv == nil {
+		return nil, fmt.Errorf(
+			"failed to cast %q object %#v to PersistentVolume", name, pvObj)
+	}
+
+	if pv.Spec.ClaimRef == nil {
+		return nil, fmt.Errorf(
+			"found PV object %q but it has a nil pv.Spec.ClaimRef indicating it is not yet bound to the claim",
+			name)
+	}
+
+	if pv.Spec.ClaimRef.UID != expectedClaimUID {
+		return nil, fmt.Errorf(
+			"found PV object %q but its pv.Spec.ClaimRef.UID (%q) does not point to claim.UID (%q)",
+			name,
+			pv.Spec.ClaimRef.UID,
+			expectedClaimUID)
+	}
+
+	// Do not return the object from the informer, since the store is shared it
+	// may be mutated by another consumer.
+	clonedPVObj, err := api.Scheme.DeepCopy(pv)
+	if err != nil || clonedPVObj == nil {
+		return nil, fmt.Errorf("failed to deep copy %q PV object", name)
+	}
+
+	clonedPV, ok := clonedPVObj.(*api.PersistentVolume)
+	if ok || clonedPV == nil {
+		return nil, fmt.Errorf(
+			"failed to cast %q clonedPV %#v to PersistentVolume", name, pvObj)
+	}
+
+	return volume.NewSpecFromPersistentVolume(clonedPV, pvcReadOnly), nil
+}
+
+// processSafeToDetachAnnotations processes the "safe to detach" annotations for
+// the given node. It makes calls to delete any annotations refering to volumes
+// it is not aware of. For volumes it is aware of, it marks them safe to detach
+// in the "actual state of world" data structure.
+func (adc *attachDetachController) processSafeToDetachAnnotations(
+	nodeName string, annotations map[string]string) {
+	var annotationsToRemove []string
+	for annotation := range annotations {
+		// Check annotations for "safe to detach" volumes
+		annotation = strings.ToLower(annotation)
+		if strings.HasPrefix(annotation, SafeToDetachAnnotation) {
+			// If volume exists in "actual state of world" mark it as safe to detach
+			safeToAttachVolume := strings.TrimPrefix(annotation, SafeToDetachAnnotation)
+			if err := adc.actualStateOfWorld.MarkVolumeNodeSafeToDetach(safeToAttachVolume, nodeName); err != nil {
+				// If volume doesn't exist in "actual state of world" remove
+				// the "safe to detach" annotation from the node
+				annotationsToRemove = append(annotationsToRemove, annotation)
+			}
+		}
+	}
+
+	// TODO: Call out to API server to delete annotationsToRemove from Node
+}
+
+// getUniquePodName returns a unique name to reference pod by in memory caches
+func getUniquePodName(pod *api.Pod) string {
```

Comment:
I'll turn it around on you then - you should have defined an opaque type
from the start.  any time you start keying on strings in objects, it's a
smell.  Consider strong types where formatted content matters.

```
type MyKey NamespacedName
```

gives you your type safety (never accidentally set the wrong format string)
and opacity.

On Tue, May 24, 2016 at 10:41 PM, Saad Ali <notifications@github.com> wrote:

> In pkg/controller/volume/attach_detach_controller.go
> <https://github.com/kubernetes/kubernetes/pull/25457#discussion_r64517837>
> :
>
> > +		if strings.HasPrefix(annotation, SafeToDetachAnnotation) {
> > +			// If volume exists in "actual state of world" mark it as safe to detach
> > +			safeToAttachVolume := strings.TrimPrefix(annotation, SafeToDetachAnnotation)
> > +			if err := adc.actualStateOfWorld.MarkVolumeNodeSafeToDetach(safeToAttachVolume, nodeName); err != nil {
> > +				// If volume doesn't exist in "actual state of world" remove
> > +				// the "safe to detach" annotation from the node
> > +				annotationsToRemove = append(annotationsToRemove, annotation)
> > +			}
> > +		}
> > +	}
> > +
> > +	// TODO: Call out to API server to delete annotationsToRemove from Node
> > +}
> > +
> > +// getUniquePodName returns a unique name to reference pod by in memory caches
> > +func getUniquePodName(pod *api.Pod) string {
>
> I'd prefer to leave the actual/desired/etc data structures keyed off a
> string rather than a struct of namespace/name. There is no reason the data
> structs need to be aware of the intricacies of what makes up a unique pod
> name. All they care about is a unique key for a given pod, that's it.
> How/what the caller (attach_detach_controller) does do generate that key is
> up to the caller. Separation of responsibilities and all that.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25457/files/6303cf96f7e05cf57ebcdcf5585ea4db5f27ed39#r64517837>
>